### ISSUE TITLE
[mmlspark] unable to allocate more than int max array

### ISSUE THREAD

**Issue by chris-smith-zocdoc** [microsoft/LightGBM#2615]
The `lightgbmlib.new_doubleArray` function in the jvm lib limits its size to integers. We should be able to allocate more memory on 64bit systems.

Is there any reason not to allow this? Does the rest of LightGBM handle large data arrays correctly?

Current signature 

```java
public static SWIGTYPE_p_double new_doubleArray(int var0)
```

Proposal: Change int to long
```java
public static SWIGTYPE_p_double new_doubleArray(long var0)
```

I'm new to SWIG so I'm not sure if their macros allow for this, but we can just write our own small functions (copies of their generated code with the types changed)

```cpp
  static double *new_doubleArray_2(int64_t nelements) {
    return new double[nelements]();
  }

  static double doubleArray_getitem_2(double *ary, int64_t index) {
      return ary[index];
  }
  static void doubleArray_setitem_2(double *ary, int64_t index, double value) {
      ary[index] = value;
  }
```

## Reproducible examples

```java
long numRows = 17000000;
int numCols = 130;
SWIGTYPE_p_double data = lightgbmlib.new_doubleArray(numCols * numRows);
```

Related mmlspark issue https://github.com/Azure/mmlspark/issues/753

**Comment by StrikerRUS** [comment]
@chris-smith-zocdoc 
> Does the rest of LightGBM handle large data arrays correctly?

I remember that **prediction** cannot handle data with more than MAX_INT rows. Support can be added in a language wrapper (#1615).

> due to the efficiency concern, LightGBM only accept the int32 nrow.
https://github.com/microsoft/LightGBM/issues/1590#issuecomment-415066108

@imatiach-msft @guolinke

**Comment by chris-smith-zocdoc** [comment]
@StrikerRUS int32 for nrows is ok in my case. The use case is millions of rows and hundreds of columns, the result of nrows * ncols exceeds int32.

In my example, 17 million rows * 130 columns gives 2,210,000,000 which overflows.

In my actual dataset I have both more rows and more columns

**Comment by guolinke** [comment]
in lightgbm side, both nrow and ncol could be int32, and the total size could be int32*int32. refer to https://github.com/microsoft/LightGBM/blob/master/src/c_api.cpp#L599-L607

ping @imatiach-msft

**Comment by imatiach-msft** [comment]
@chris-smith-zocdoc @guolinke thank you for finding this issue, indeed it looks like a bug in the wrapper logic.  I will look into fixing it as soon as I have some free cycles.

**Comment by imatiach-msft** [comment]
@chris-smith-zocdoc  @guolinke it looks like the carrays.i swig macros only support integer arguments:
https://github.com/swig/swig/blob/master/Lib/carrays.i#L23
so we can probably just define our own macro and use that.  Another option might be to try and enhance the SWIG library.  For now I opened an issue in the SWIG library, I'm curious to see what they say:

https://github.com/swig/swig/issues/1680

**Comment by chris-smith-zocdoc** [comment]
I think including our own methods is fine, I don't think we need to write our own macro since its just three small methods that need to be fixed. The code I included in the issue description will generate the java methods with a `long` parameter

**Comment by StrikerRUS** [comment]
Closed via #2859.
